File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>
#include <iostream>
#include <queue>

using namespace std;


int answer = 0;
// d = 1 or -1  => sum += d * num
void dfs(vector<int> numbers, int target, int sum, int idx){
    if (idx == numbers.size()){
        if (sum == target){
            answer ++;
        }
        return;
    }
    
    dfs(numbers, target, sum + numbers[idx], idx+1);
    dfs(numbers, target, sum - numbers[idx], idx+1);
}

int solution(vector<int> numbers, int target) {
    dfs(numbers, target, 0, 0);
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    public int solution(int[] numbers, int target) {
        int currSum = 0;
        return search(1, currSum + numbers[0], numbers, target) + search(1, currSum - numbers[0], numbers, target);
    }
    
    public int search(int idx, int currSum, int[] numbers, int target){
        if (idx == numbers.length){
            if (currSum == target) return 1;
            else return 0;
        }
        return search(idx + 1, currSum + numbers[idx], numbers, target) + search(idx + 1, currSum - numbers[idx], numbers, target);       
    }
}